sklearn/metrics/tests/test_common.py

```diff
@@ -1110,3 +1112,97 @@ def test_no_averaging_labels():
             score_labels = metric(y_true, y_pred, labels=labels, average=None)
             score = metric(y_true, y_pred, average=None)
             assert_array_equal(score_labels, score[inverse_labels])
+
+
+def test_metric_permutation_invariance_multiclass():
+    rng = np.random.RandomState(42)
+    y_true = rng.randint(0, 20, 100)
+    y_pred = rng.randint(0, 20, 100)
+    classes_perm = rng.permutation(20)
+    y_true_perm = classes_perm[y_true]
+    y_pred_perm = classes_perm[y_pred]
+    for name in CLASSIFICATION_METRICS:
+        if name in METRIC_UNDEFINED_BINARY_MULTICLASS:
+            continue
+        if name is "confusion_matrix":
+            continue
+        metric = ALL_METRICS[name]
+        score = metric(y_true, y_pred)
+        score_perm = metric(y_true_perm, y_pred_perm)
+        assert_almost_equal(score, score_perm)
+
+
+def test_metric_permutation_invariance_multilabel():
+    y_true = []
+    y_pred = []
+    rng = np.random.RandomState(42)
+    for i in range(0, 100):
+        n_labels_true = rng.randint(1, 10)
+        n_labels_pred = rng.randint(1, 10)
+        y_true.append(rng.randint(0, 20, n_labels_true))
+        y_pred.append(rng.randint(0, 20, n_labels_pred))
+    classes_perm = rng.permutation(20)
+    y_true_perm = MultiLabelBinarizer().fit_transform([classes_perm[y]
```

Comment:
You should have no need for `MultiLabelBinarizer`.

You could just generate correlated binary `y_true` and `y_pred` with
```python
R = np.random.rand(n_samples, n_classes)
y_true = R > p
y_pred = R + np.random.randn(n_samples, n_classes) > p
```
p = .7, for instance. Requiring correlation means that the evaluation is interesting.

Permutation then just requires `y[:, classes_perm]`, I think.